sdks/python/apache_beam/runners/worker/sdk_worker.py

```diff
@@ -38,82 +37,147 @@
-      # WhichOneOf returns the name of the set field as a single string
-      if request_type in ['process_bundle_progress']:
-        thread_pool = self._progress_thread_pool
-      else:
-        thread_pool = self._default_work_thread_pool
-
-      # Need this wrapper to capture the original stack trace.
-      def do_instruction(request):
-        try:
-          return self.worker.do_instruction(request)
-        except Exception as e:  # pylint: disable=broad-except
-          traceback_str = traceback.format_exc(e)
-          raise Exception("Error processing request. Original traceback "
-                          "is\n%s\n" % traceback_str)
-
-      def handle_response(request, response_future):
-        try:
-          response = response_future.result()
-        except Exception as e:  # pylint: disable=broad-except
-          logging.error(
-              'Error processing instruction %s',
-              request.instruction_id,
-              exc_info=True)
-          response = beam_fn_api_pb2.InstructionResponse(
-              instruction_id=request.instruction_id,
-              error=str(e))
-        responses.put(response)
-
-      thread_pool.submit(do_instruction, work_request).add_done_callback(
-          functools.partial(handle_response, work_request))
-
-    logging.info("No more requests from control plane")
-    logging.info("SDK Harness waiting for in-flight requests to complete")
+      # Name spacing the request method with 'request_'. The called method
+      # will be like self.request_register(request)
+      getattr(self, SdkHarness.REQUEST_METHOD_PREFIX + request_type)(
+          work_request)
+
+    logging.info('No more requests from control plane')
+    logging.info('SDK Harness waiting for in-flight requests to complete')
     # Wait until existing requests are processed.
     self._progress_thread_pool.shutdown()
-    self._default_work_thread_pool.shutdown()
+    self._process_thread_pool.shutdown()
     # get_responses may be blocked on responses.get(), but we need to return
     # control to its caller.
-    responses.put(no_more_work)
+    self._responses.put(no_more_work)
     self._data_channel_factory.close()
-    state_handler.done()
+    # Stop all the workers and clean all the associated resources
+    for worker in self.workers.queue:
+      worker.state_handler.done()
     logging.info('Done consuming work.')
 
+  def _execute(self, task, request):
+    try:
+      response = task()
+    except Exception as e:  # pylint: disable=broad-except
+      traceback.print_exc(file=sys.stderr)
+      logging.error(
+          'Error processing instruction %s. Original traceback is\n%s\n',
+          request.instruction_id,
+          traceback.format_exc(e),
+          exc_info=True)
+      response = beam_fn_api_pb2.InstructionResponse(
+          instruction_id=request.instruction_id, error=str(e))
+    self._responses.put(response)
+
+  def _request_register(self, request):
```

Comment:
Can you add a comment here about, why we do not need a lock to modify `self._fns`